epoxy-processor/src/main/java/com/airbnb/epoxy/ClassToGenerateInfo.java

```diff
@@ -81,26 +90,39 @@ private void collectOriginalClassConstructors(TypeElement originalClass) {
    * duplicate them in the generated class for chaining purposes
    */
   private void collectMethodsReturningClassType(TypeElement originalClass) {
-    for (Element subElement : originalClass.getEnclosedElements()) {
-      Set<Modifier> modifiers = subElement.getModifiers();
-      if (subElement.getKind() == ElementKind.METHOD
-          && ((ExecutableType) subElement.asType()).getReturnType().equals(originalClass.asType())
-          && !modifiers.contains(Modifier.PRIVATE)
-          && !modifiers.contains(Modifier.STATIC)) {
-        List<? extends TypeMirror> params = ((ExecutableType) subElement.asType())
-            .getParameterTypes();
-        String methodName = subElement.getSimpleName().toString();
-        if (params.size() == 1) {
-          TypeMirror param = params.get(0);
-          ParameterSpec parameterSpec = ParameterSpec.builder(TypeName.get(param),
-              methodName).build();
-          methodsReturningClassType.add(new MethodInfo(methodName, modifiers,
-              Collections.singletonList(parameterSpec)));
-        } else {
-          methodsReturningClassType.add(new MethodInfo(methodName, modifiers,
-              buildParamList(params)));
+    TypeElement clazz = originalClass;
+    while (clazz.getSuperclass().getKind() != TypeKind.NONE) {
+      for (Element subElement : clazz.getEnclosedElements()) {
+        Set<Modifier> modifiers = subElement.getModifiers();
+        if (subElement.getKind() == ElementKind.METHOD
+            && !modifiers.contains(Modifier.PRIVATE)
+            && !modifiers.contains(Modifier.STATIC)) {
+          TypeMirror methodReturnType = ((ExecutableType) subElement.asType()).getReturnType();
+          if (methodReturnType.equals(clazz.asType())
+              || typeUtils.isSubtype(clazz.asType(), methodReturnType)) {
+            List<? extends TypeMirror> params = ((ExecutableType) subElement.asType())
+                .getParameterTypes();
+            String methodName = subElement.getSimpleName().toString();
+            if (params.size() == 1) {
+              TypeMirror param = params.get(0);
+              ParameterSpec parameterSpec;
+              if (methodName.equals(LAYOUT_METHOD) && param.getKind() == TypeKind.INT) {
```

Comment:
adding a brief comment here about why we need to handle layout specially would probably be helpful in the future